src/System.Linq.Expressions/tests/Dynamic/DynamicObjectTests.cs

```diff
@@ -0,0 +1,791 @@
+
+                        break;
+                    case 2:
+                        if (indexes[0] is int && indexes[1] is int)
+                        {
+                            string value = _twoDimensions[Tuple.Create((int)indexes[0], (int)indexes[1])];
+                            result = value;
+                            return true;
+                        }
+
+                        break;
+                }
+
+                result = null;
+                return false;
+            }
+
+            public override bool TrySetIndex(SetIndexBinder binder, object[] indexes, object value)
+            {
+                switch (indexes.Length)
+                {
+                    case 1:
+                        if (indexes[0] is int && value is int)
+                        {
+                            _oneDimension[(int)indexes[0]] = (int)value;
+                            return true;
+                        }
+
+                        break;
+                    case 2:
+                        if (indexes[0] is int && indexes[1] is int && value is string)
+                        {
+                            _twoDimensions[Tuple.Create((int)indexes[0], (int)indexes[1])] = (string)value;
+                            return true;
+                        }
+
+                        break;
+                }
+
+                return false;
+            }
+        }
+
+        private class NegatableNum : DynamicObject
+        {
+            public NegatableNum(int value)
+            {
+                Value = value;
+            }
+
+            public int Value { get; }
+
+            public override bool TryUnaryOperation(UnaryOperationBinder binder, out object result)
+            {
+                if (binder.Operation == ExpressionType.Negate)
+                {
+                    result = new NegatableNum(unchecked(-Value));
+                    return true;
+                }
+
+                result = null;
+                return false;
+            }
+        }
+
+        private class AddableNum : DynamicObject
+        {
+            public AddableNum(int value)
+            {
+                Value = value;
+            }
+
+            public int Value { get; }
+
+            public override bool TryBinaryOperation(BinaryOperationBinder binder, object arg, out object result)
+            {
+                if (binder.Operation == ExpressionType.Add)
+                {
+                    AddableNum addend = arg as AddableNum;
+                    if (addend != null)
+                    {
+                        result = new AddableNum(Value + addend.Value);
+                        return true;
+                    }
+                }
+
+                result = null;
+                return false;
+            }
+        }
+
+        private class Accumulator : DynamicObject
+        {
+            // Allows addition to with this on the lhs, but not with it on the rhs
```

Comment:
I was using "this" in the C# sense of "itself", but no, it doesn't seem very readable on second glance.